src/Microsoft.AspNetCore.Routing/Matching/RefEmitTrieJumpTable.cs

```diff
@@ -0,0 +1,167 @@
+
+            var nextGroup = il.DefineLabel();
+            var groups = _entries.GroupBy(e => e.text.Length).ToArray();
+            for (var i = 0; i < groups.Length; i++)
+            {
+                var group = groups[i];
+
+                // Similar to 'if (length != X) { ... }
+                il.MarkLabel(nextGroup);
+                nextGroup = il.DefineLabel();
+
+                il.Emit(OpCodes.Ldarg_2);
+                il.Emit(OpCodes.Ldc_I4, group.Key);
+                il.Emit(OpCodes.Beq, nextGroup);
+
+                // Process the group
+                EmitTableForStringsOfEqualLength(il, group.ToArray(), 0, group.Key, offset, c, returnDefault);
+            }
+
+            il.MarkLabel(nextGroup);
+            il.MarkLabel(returnDefault);
+            il.Emit(OpCodes.Ldc_I4, _defaultDestination);
+            il.Emit(OpCodes.Ret);
+
+#if NET462
+            var t = type.CreateType();
+            assembly.Save(assemblyName.Name + ".dll");
+            return (Func<string, int, int, int>)t.GetMethod("GeneratedGetDestination").CreateDelegate(typeof(Func<string, int, int, int>));
+#else
+            return (Func<string, int, int, int>)method.CreateDelegate(typeof(Func<string, int, int, int>));
+#endif
+        }
+
+        private static void EmitTableForStringsOfEqualLength(
+            ILGenerator il,
+            (string text, int destination)[] entries,
+            int index,
+            int length,
+            LocalBuilder offset,
+            LocalBuilder c,
+            Label returnDefault)
+        {
+            // We've reached the end of the string.
+            if (index == length)
+            {
+                Debug.Assert(entries.Length == 1, "We should have a single entry");
+                il.Emit(OpCodes.Ldc_I4, entries[0].destination);
+                il.Emit(OpCodes.Ret);
+                return;
+            }
+
+            // offset++
+            il.Emit(OpCodes.Ldloc, offset);
+            il.Emit(OpCodes.Ldc_I4, 1);
+            il.Emit(OpCodes.Add);
+            il.Emit(OpCodes.Stloc, offset);
+
+            // c = path[offset]
+            il.Emit(OpCodes.Ldarg_1);
+            il.Emit(OpCodes.Ldloc, offset);
+            il.Emit(OpCodes.Callvirt, typeof(string).GetProperty("Chars", new Type[] { typeof(int), }).GetMethod);
+            il.Emit(OpCodes.Stloc, c);
+
+            var nextGroup = il.DefineLabel();
+            var groups = entries.GroupBy(e => char.ToUpperInvariant(e.text[index])).ToArray();
+            for (var i = 0; i < groups.Length; i++)
+            {
+                var group = groups[i];
+
+                // Similar to 'if (c == 'A' || c == 'a') { ... }
+                il.MarkLabel(nextGroup);
+                nextGroup = il.DefineLabel();
+
+                il.Emit(OpCodes.Ldloc, c);
+                il.Emit(OpCodes.Ldc_I4, group.Key);
+                il.Emit(OpCodes.Beq, nextGroup);
+
+                if (char.ToLowerInvariant(group.Key) != group.Key)
```

Comment:
To not pay the cost of a case insensitive comparison of the char if the does doesn't have multiple cases? Very clever 😁

Reply:
It will get better - https://matthewarcus.wordpress.com/2014/02/12/tolower/